README.md
# vgwrap

[![Latest Version on Packagist][ico-version]][link-packagist]
[![Software License][ico-license]](LICENSE.md)
[![StyleCI][ico-styleci]][link-styleci]
[![Total Downloads][ico-downloads]][link-downloads]

A Laravel 5 wrapper for the Vainglory API

## Structure

This project's structure.

```
src/
    config/
    Facade.php
    ServiceProvider.php
    VGWrapClient.php
    VGWrapException.php
tests/
...
```

## Install

Via Composer

``` bash
$ composer require agangofkittens/vgwrap
```

## Configuration

In your `config/app.php` add `agangofkittens\vgwrap\ServiceProvider::class,` to the end of providers array.
``` php
'providers' => [
    App\Providers\RouteServiceProvider::class,
    ...
    'agangofkittens\vgwrap\ServiceProvider',
],
```
At the end of `config/app.php` add `'VGWrap' => agangofkittens\vgwrap\Facade::class,` to the aliases array.
``` php
'aliases' => [
    'App' => Illuminate\Support\Facades\App::class,
    ...
    'VGWrap' => agangofkittens\vgwrap\Facade::class,
],
```

Next publish the config file:
``` bash
$ php artisan vendor:publish
```

Finally set your [API key](https://developer.vainglorygame.com/) by updating the `api-id` value in `config/vgwrap.php` or your `.env` file.
If necessary you can also modify the API endpoints in the aforementioned `config/vgwrap.php` file.

## Usage

``` php
// Fetch a player by using a playerName filter
$response = \VGWrap::get('shards/eu/players', [
    'query' => [
        'filter' => [
            'playerNames' => 'Adrame'
        ]
    ]
]);
$player = json_decode($response->getBody()->getContents());

// Or using a shortcut
$response = \VGWrap::getData('shards/eu/players', [
    'query' => [
        'filter' => [
            'playerNames' => 'Adrame'
        ]
    ]
]);
```

_For more examples and usage, please refer to the [documentation][link-docs]._

## Change log

Please see [CHANGELOG](CHANGELOG.md) for more information on what has changed recently.

## Testing

``` bash
$ composer test
```

## Contributing

Please see [CONTRIBUTING](CONTRIBUTING.md) and [CONDUCT](CONDUCT.md) for details.

## Security

If you discover any security related issues, please email agangofkittens@gmail.com instead of using the issue tracker.

## Credits

- [agangofkittens][link-author]
- [All Contributors][link-contributors]

## License

The MIT License (MIT). Please see [License File](LICENSE.md) for more information.

[ico-version]: https://img.shields.io/packagist/v/agangofkittens/vgwrap.svg?style=flat-square
[ico-license]: https://img.shields.io/badge/license-MIT-brightgreen.svg?style=flat-square
[ico-downloads]: https://img.shields.io/packagist/dt/agangofkittens/vgwrap.svg?style=flat-square
[ico-styleci]: https://styleci.io/repos/98232070/shield

[link-packagist]: https://packagist.org/packages/agangofkittens/vgwrap
[link-downloads]: https://packagist.org/packages/agangofkittens/vgwrap
[link-author]: https://github.com/agangofkittens
[link-contributors]: ../../contributors
[link-styleci]: https://styleci.io/repos/98232070
[link-docs]: https://vgwrap.emillie.ml


CHANGELOG.md
# Changelog

All Notable changes to `vgwrap` will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [1.0.2] - 2017-07-26
- Clean up VGWrapClient.php
- Fix the base-uri since API endpoint isn't limited to `/shards/`.

## [1.0.1] - 2017-07-26
### Added
- Documentation fixes.
- Add to packagist.

## 1.0.0 - 2017-07-26
- First public release.
- Fixed some mistakes in the README.md file.

[Unreleased]: https://github.com/agangofkittens/vgwrap/compare/1.0.2...HEAD
[1.0.2]: https://github.com/agangofkittens/vgwrap/compare/1.0.1...1.0.2
[1.0.1]: https://github.com/agangofkittens/vgwrap/compare/1.0.0...1.0.1

tests/ClientTest.php
<?php
/**
 * Created by PhpStorm.
 * User: ashe
 * Date: 7/24/17
 * Time: 11:44 PM
 */

use agangofkittens\vgwrap\VGWrapClient;
use GuzzleHttp\Client as GuzzleClient;
use PHPUnit\Framework\TestCase;

/**
 * Class ClientTest
 *
 * @author agangofkittens <agangofkittens@gmail.com>
 */
class ClientTest extends TestCase
{
    public function testCreateDefaultClient()
    {
        $client = new VGWrapClient('apikey');
        $this->assertInstanceOf('GuzzleHttp\\ClientInterface', $client->getClient());
    }

    public function testUseCustomClient()
    {
        $guzzle = new GuzzleClient();
        $client = new VGWrapClient('apikey');
        $this->assertSame($client, $client->setClient($guzzle));
        $this->assertSame($guzzle, $client->getClient());
    }

    public function testAPIkey()
    {
        $client = new VGWrapClient('apikey');
        $this->assertSame('apikey', $client->getAPIkey());
    }
}


src/VGWrapClient.php
<?php
/**
 * Created by PhpStorm.
 * User: ashe
 * Date: 7/24/17
 * Time: 7:19 PM
 */

namespace agangofkittens\vgwrap;

use GuzzleHttp\Client;
use GuzzleHttp\ClientInterface;
use Psr\Http\Message\ResponseInterface;

/**
 * Class VGWrapClient
 *
 * Simple PHP client to interact with the Vainglory API
 *
 * @method ResponseInterface get($uri, array $options = [])
 * @method ResponseInterface head($uri, array $options = [])
 * @method ResponseInterface put($uri, array $options = [])
 * @method ResponseInterface post($uri, array $options = [])
 * @method ResponseInterface patch($uri, array $options = [])
 * @method ResponseInterface delete($uri, array $options = [])
 *
 * @package agangofkittens\vgwrap
 * @author  agangofkittens <agangofktitens@gmail.com>
 * @version 1.0.3
 * @access  public
 * @see     https://vgwrap.emillie.ml
 */
class VGWrapClient
{
    /**
     * @var object \GuzzleHttp\Client
     */
    private $client;

    /**
     * @var string api_key
     */
    protected $api_key;

    /**
     * @var string base_uri
     */
    protected $base_uri;

    /**
     * Constructor method for VGWrapClient class
     *
     * @param $api_key
     * @param array
     */
    public function __construct($api_key, $base_uri)
    {
        $this->setAPIkey($api_key);
        $this->setBaseUri($base_uri);
    }

    /**
     * Set the API key
     *
     * @param $api_key
     * @return mixed
     */
    public function setAPIkey($api_key)
    {
        return ($this->api_key = $api_key);
    }

    /**
     * Get the API key
     *
     * @return string API_KEY
     */
    public function getAPIkey()
    {
        return $this->api_key;
    }

    /**
     * Set base URL
     *
     * @param $base_uri
     * @return mixed
     */
    public function setBaseUri($base_uri)
    {
        return ($this->base_uri = $base_uri);
    }

    /**
     * Get base URL
     *
     * @return mixed
     */
    public function getBaseUri()
    {
        return $this->base_uri;
    }

    /**
     * Set client
     *
     * @param ClientInterface $client
     * @return $this
     */
    public function setClient(ClientInterface $client)
    {
        $this->client = $client;
        return $this;
    }

    /**
     * Get GuzzleClient, create if it's null
     *
     * @param array $config
     * @return Client
     */
    public function getClient(array $config = [])
    {
        $config = array_merge($config, [
            'base_uri' => $this->getBaseUri(),
            'headers' => [
                'Authorization' => $this->getAPIkey(),
                'Accept' => 'application/vnd.api+json',
                'Accept-Encoding' => 'gzip'
            ]

        ]);

        if (!$this->client) {
            $this->client = new Client($config);
        }

        return $this->client;
    }

    /**
     * Shortcut
     *
     * @param $uri
     * @param array $options
     * @return mixed
     */
    public function getData($uri, $options = [])
    {
        return json_decode($this->getClient()->get($uri, $options)->getBody()->getContents());
    }

    /**
     * Perform a request
     *
     * @param $method
     * @param string $uri
     * @param array $options
     * @return mixed|ResponseInterface
     */
    public function request($method, $uri = '', array $options = [])
    {
        return $this->client->request($method, $uri, $options);
    }

    /**
     * Forward any other call to guzzle client.
     *
     * @param $method
     * @param $parameters
     * @return mixed
     */
    public function __call($method, $parameters)
    {
        return call_user_func_array([$this->getClient(), $method], $parameters);
    }
}


src/ServiceProvider.php
<?php
/**
 * Created by PhpStorm.
 * User: ashe
 * Date: 7/24/17
 * Time: 7:08 PM
 */

namespace agangofkittens\vgwrap;

use Illuminate\Support\ServiceProvider as IlluminateServiceProvider;
use Config;

/**
 * Class ServiceProvider
 *
 * @author agangofkittens
 * @package agangofkittens\vgwrap
 */
class ServiceProvider extends IlluminateServiceProvider
{
    /**
     * Indicates if loading of the provider is deferred.
     *
     * @var bool
     */
    protected $defer = false;

    /**
     * Bootstrap the application events.
     *
     * @return void
     */
    public function boot()
    {
        $configPath = __DIR__ . '/config/config.php';

        if (function_exists('config_path')) {
            $publishPath = config_path('vgwrap.php');
        } else {
            $publishPath = base_path('config/vgwrap.php');
        }

        // Publish config files
        $this->publishes([$configPath => $publishPath], 'config');
    }

    /**
     * Register the service provider.
     *
     * @return void
     */
    public function register()
    {
        $configPath = __DIR__ . '/config/config.php';

        $this->app->bind('vgwrap', function ($app) {
            return new VGWrapClient(config('vgwrap.api-key'), config('vgwrap.base-uri'));
        });
    }

    /**
     * Get the services provided by the provider.
     *
     * @return array
     */
    public function provides()
    {
        return array('vgwrap');
    }
}
